[Doc] Update the dynamo deepdive doc (#108147)
pytorch/pytorch: docs/source/torch.compiler_deepdive.rst

How to inspect artifacts generated by TorchDynamo?
--------------------------------------------------

To inspect the artifacts generated by TorchDynamo, there is an API ``torch._dynamo.eval_frame._debug_get_cache_entry_list`` that retrieves compiled code and guards out of a function's ``__code__`` object. A compiled function can have several cache entries, and each cache entry consists a generated function to check guards, and a ``types.CodeType`` object to keep the code to be executed if the guarding conditions are satisfied.

.. code-block:: python

   from torch._dynamo.eval_frame import _debug_get_cache_entry_list
   cache_entries = _debug_get_cache_entry_list(toy_example._torchdynamo_orig_callable.__code__)
   guard, code = cache_entries[0]
   # the guard takes the local variables of an input frame, and tells whether a re-compilation should be triggered.
   import dis
   dis.dis(guard)
   dis.dis(code)

> If you know python bytecode, you can understand the above output. There is also a nice tool "depyf" to convert the bytecode into human readable source code. To run the following code, please first run "pip install depyf".

Reviewer: ```suggestion
If you know python bytecode, you can understand the above output. There is also a tool "depyf" to convert the bytecode into human readable source code. To run the following code, please first run ``pip install depyf``.
```

.. code-block:: python

   from depyf import decompile
   print("guard code:")
   print(decompile(guard))
   print("compiled code:")
   print(decompile(code))

> The output is something like the following:

Reviewer: ```suggestion
The output is:
```

The graph break is shown in the code of compiled ``toy_example``, where we have to use python interpreter to select the following graph to execute.

Note that we pass a simple ``my_compiler`` function as the backend compiler, therefore the subgraph code ``__resume_at_38_2``, ``__resume_at_30_1``, and ``__compiled_fn_0`` remain python code. This can also be inspected (please ignore the function name, and only use the function signature and function body code):

.. code-block:: python

   import inspect
   print("source code of __compiled_fn_0:")
   print(inspect.getsource(__compiled_fn_0))
   print("source code of __resume_at_30_1:")
   print(inspect.getsource(__resume_at_30_1))
   print("source code of __resume_at_38_2:")
   print(inspect.getsource(__resume_at_38_2))

::

   source code of __compiled_fn_0:
   def ignore_this_function_name(self, L_a_ : torch.Tensor, L_b_ : torch.Tensor):
       l_a_ = L_a_
       l_b_ = L_b_
       abs_1 = torch.abs(l_a_)
       add = abs_1 + 1;  abs_1 = None
       truediv = l_a_ / add;  l_a_ = add = None
       sum_1 = l_b_.sum();  l_b_ = None
       lt = sum_1 < 0;  sum_1 = None

>        return (truediv, lt)

Reviewer: This is actually an FX graph (since the compiler above returned the original graph), so if you just call `print()` on it it might be cleaner than getsource().

Author: This is not FX graph actually. It is the forward function of that graph. `print` on it gives `<function forward at 0x17c7212d0>`

Reviewer: Ah, you could print(`fn.__self__`) then.  Or change the code above to not return .forward.

Author: Okay, somewhat complicated, but `print(__compiled_fn_0._torchdynamo_orig_callable.__self__)` does work.

   source code of __resume_at_30_1:
   def ignore_this_function_name(a, b):
       x = a / (torch.abs(a) + 1)
       if b.sum() < 0:
           b = b * -1
       return x * b

   source code of __resume_at_38_2:
   def ignore_this_function_name(a, b):
       x = a / (torch.abs(a) + 1)
       if b.sum() < 0:
           b = b * -1

>        return x * b

Reviewer: This source code is wrong.  It is just showing the original code, without any bytecode level changes done by dynamo.

If you print out the bytecode you will see:
1) different args to the function not in this source code
2) `JUMP_ABSOLUTE` as the first instructions, without any corresponding line in the source code.  This causes these functions to start in the middle rather than at the top.

In reality, it is more like:
```
 def __resume_at_30_1(b, x):
       JUMP_ABSOLUTE <target>
       x = a / (torch.abs(a) + 1)
       if b.sum() < 0:
           <target>
           b = b * -1
       return x * b
```

There is actually no way to represent this precisely with Python source code, since python doesn't have a `goto` instruction.

Author: No worry, this can be easily fixed by removing unreachable bytecode.

Author: I fixed it in [4678346](https://github.com/pytorch/pytorch/pull/108147/commits/4678346e1414da1fc2fe9c3868d7efaca25614f0). Previously, `inspect.getsource` does not work for `__resume_at_xxx` functions. Now I use `depyf.decompile` to decompile its source code.

Meanwhile, I noticed that the function names of `__resume_at_xxx` are not valid python function names. Maybe this should be fixed in pytorch master? WDYT?

Reviewer: What makes it invalid?
```
>>> def __resume_at_30_1(b, x):
...   pass
... 
>>> 
```

Author: `__resume_at_30_1.__code__.co_name` is actually `<resume in toy_example>`

Reviewer: ah, that is for stack traces

Author: Possible fix: change https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/resume_execution.py#L368 this line to a valid variable name.

A better fix: change it to be the same as the function variable name, i.e. `__resume_at_30_1 ` stuff.

How to inspect artifacts generated by TorchDynamo?
--------------------------------------------------

To inspect the artifacts generated by TorchDynamo, there is an API ``torch._dynamo.eval_frame._debug_get_cache_entry_list`` that retrieves compiled code and guards out of a function's ``__code__`` object. A compiled function can have several cache entries, and each cache entry consists a generated function to check guards, and a ``types.CodeType`` object to keep the code to be executed if the guarding conditions are satisfied.

.. code-block:: python

   from torch._dynamo.eval_frame import _debug_get_cache_entry_list
   cache_entries = _debug_get_cache_entry_list(toy_example._torchdynamo_orig_callable.__code__)
   guard, code = cache_entries[0]
   # the guard takes the local variables of an input frame, and tells whether a re-compilation should be triggered.
   import dis
   dis.dis(guard)
   dis.dis(code)

> If you know python bytecode, you can understand the above output. There is also a tool "depyf" to convert the bytecode into human readable source code. To run the following code, please first run ``pip install depyf``.

Reviewer 2: ```suggestion
If you know Python bytecode, you can understand the above output. There is also a tool ``depyf`` to convert the bytecode into human-readable source code. If you don't have ``depyf`` already installed, run ``pip install depyf`` before running the code below.
```

       if not getattr(___guarded_code, 'valid'):
           return False
       _var0 = L['a']
       if not hasattr(_var0, '_dynamo_dynamic_indices') == False:
           return False
       _var1 = L['b']
       if not hasattr(_var1, '_dynamo_dynamic_indices') == False:
           return False
       if not ___is_grad_enabled():
           return False
       if ___are_deterministic_algorithms_enabled():
           return False
       if not ___is_torch_function_enabled():
           return False
       if not getattr(utils_device, 'CURRENT_DEVICE') == None:
           return False
       if not ___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names
           ):
           return False
       return True

   compiled code:
   def toy_example(a, b):
       __temp_1 = __compiled_fn_0(a, b)
       x = __temp_1[0]
       if __temp_1[1]:
           return __resume_at_30_1(b, x)
       return __resume_at_38_2(b, x)

Some names referenced in the code are:

- Compiled functions, stored in the global namespace of the module containing the original function ``toy_example``. These include names like ``__compiled_fn_0`` / ``__resume_at_30_1`` / ``__resume_at_38_2``.

> - Clousure variables used for checking guards. The names can be accessed from ``guard.__code__.co_freevars``, and the values are stored in ``guard.__closure__``. These include names like ``___guarded_code`` / ``___is_grad_enabled`` / ``___are_deterministic_algorithms_enabled`` / ``___is_torch_function_enabled`` / ``utils_device`` / ``___check_tensors`` / ``tensor_check_names``.

Reviewer 2: ```suggestion
- Closure variables used for checking guards. The names can be accessed from ``guard.__code__.co_freevars``, and the values are stored in ``guard.__closure__``. These include names like ``___guarded_code`` / ``___is_grad_enabled`` / ``___are_deterministic_algorithms_enabled`` / ``___is_torch_function_enabled`` / ``utils_device`` / ``___check_tensors`` / ``tensor_check_names``.
```

- Argument ``L`` of the ``guard`` function. This is a dict mapping the name of arguments of ``toy_example`` to its values. This is only available when the function is called, where the frame evaluation API comes into play. In short, ``L`` is a ``dict`` with structure of ``{'a': value_a, 'b': value_b}``. Therefore, you can see the code uses ``L['a']`` to refer to the input variable ``a``.

> The graph break is shown in the code of compiled ``toy_example``, where we have to use python interpreter to select the following graph to execute.

Reviewer 2: ```suggestion
The graph break is shown in the code of compiled ``toy_example``, where we have to use Python interpreter to select the following graph to execute.
```

> Note that we pass a simple ``my_compiler`` function as the backend compiler, therefore the subgraph code ``__resume_at_38_2``, ``__resume_at_30_1``, and ``__compiled_fn_0`` remain python code. This can also be inspected (please ignore the function name, and only use the function signature and function body code):

Reviewer 2: ```suggestion
Note that we pass a simple ``my_compiler`` function as the backend compiler, therefore the subgraph code ``__resume_at_38_2``, ``__resume_at_30_1``, and ``__compiled_fn_0`` remain Python code. This can also be inspected (please ignore the function name, and only use the function signature and function body code):
```

.. code-block:: python

   print("source code of __compiled_fn_0:")
   print(__compiled_fn_0._torchdynamo_orig_callable.__self__)
   print("=" * 60)
   print("source code of __resume_at_30_1:")
   print(decompile(__resume_at_30_1))
   print("=" * 60)
   print("source code of __resume_at_38_2:")
   print(decompile(__resume_at_38_2))

::

   source code of __compiled_fn_0:
   GraphModule()

   def forward(self, L_a_ : torch.Tensor, L_b_ : torch.Tensor):
       l_a_ = L_a_
       l_b_ = L_b_
       abs_1 = torch.abs(l_a_)
       add = abs_1 + 1;  abs_1 = None
       truediv = l_a_ / add;  l_a_ = add = None
       sum_1 = l_b_.sum();  l_b_ = None
       lt = sum_1 < 0;  sum_1 = None
       return (truediv, lt)

>    # To see more debug info, please use `graph_module.print_readable()`

Reviewer 2: ```suggestion
   # To see more debug info, please use ``graph_module.print_readable()``
```

   ============================================================
   source code of __resume_at_30_1:
   def <resume in toy_example>(b, x):
       b = b * -1
       return x * b

   ============================================================
   source code of __resume_at_38_2:
   def <resume in toy_example>(b, x):
       return x * b

> However, keep in mind that, if we use other backends like the built-in `inductor`, the subgraph code will be compiled CUDA kernels for GPU or C++ code for CPU.

Reviewer 2: ```suggestion
However, if we use other backends like the built-in ``inductor``, the subgraph code will be compiled CUDA kernels for GPU or C++ code for CPU.
```

To summarize, the compiled code is conceptually equivalent to the code below:

.. code-block:: python

   def compiled_example(a, b):
       L = {'a': a, 'b': b}
       for guard, code in get_cache_entries():
           if guard(L):
               return code(a, b)
       recompile_and_add_another_cache_entry()

The following diagram demonstrates how ``torch.compile`` transforms and optimizes user-written code:

> .. image:: _static/img/dynamo/flowchart.jpg

Reviewer 2: Is it possible to add a paragraph that describes what is going on on the diagram?

Author: Sure, fixed in [454676e](https://github.com/pytorch/pytorch/pull/108147/commits/454676e32916ce8381e10a111cf98d589cb879ac).
